<a href="https://colab.research.google.com/github/mmilannaik/BigOCheatSheet/blob/master/MA_AB_testing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [12]:
import pandas as pd
import numpy as np
from numpy import percentile
import warnings
warnings.filterwarnings('ignore')

import scipy.stats as stats
from scipy.stats import shapiro

In [2]:
df = pd.read_csv('/content/ab_data.csv')
df.head()

user_id timestamp      group landing_page  converted
0   851104   11:48.6    control     old_page          0
1   804228   01:45.2    control     old_page          0
2   661590   55:06.2  treatment     new_page          0
3   853541   28:03.1  treatment     new_page          0
4   864975   52:26.2    control     old_page          1

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 294480 entries, 0 to 294479
Data columns (total 5 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       294480 non-null  int64 
 1   timestamp     294480 non-null  object
 2   group         294480 non-null  object
 3   landing_page  294480 non-null  object
 4   converted     294480 non-null  int64 
dtypes: int64(2), object(3)
memory usage: 11.2+ MB


In [4]:
df[df['user_id'].duplicated()]

user_id timestamp      group landing_page  converted
2656     698120   13:42.6    control     old_page          0
2893     773192   55:59.6  treatment     new_page          0
7500     899953   06:54.1    control     new_page          0
8036     790934   32:20.3  treatment     new_page          0
10218    633793   16:00.7  treatment     old_page          0
...         ...       ...        ...          ...        ...
294309   787083   15:21.0    control     old_page          0
294328   641570   59:27.7    control     old_page          0
294331   689637   34:28.3    control     new_page          0
294355   744456   32:07.1  treatment     new_page          0
294478   759899   20:29.0  treatment     new_page          0

[3895 rows x 5 columns]

In [5]:
# Check if there is mismatch between group and landing_page
df_mismatch = df[(df["group"]=="treatment")&(df["landing_page"]=="old_page")
                |(df["group"]=="control")&(df["landing_page"]=="new_page")]
n_mismatch = df_mismatch.shape[0]
print(f"The number of mismatched rows:{n_mismatch} rows" )
print("Percent of mismatched rows:%.2f%%" % (n_mismatch/df.shape[0]*100))

The number of mismatched rows:3893 rows
Percent of mismatched rows:1.32%


As you can see, there are 3893 rows mismatched, which means that treatment group mismatched with old_page and control group mismatched with new_page.There are also 3895 duplicated user_id, some of them loading both new page and old page. This might be the relative department repeat the expermient after finding the mismatched problem. Therefore, we need to only keep the right data and drop the mismatched data.

In [6]:
df2 = df[(df["group"]=="treatment")&(df["landing_page"]=="new_page")
                |(df["group"]=="control")&(df["landing_page"]=="old_page")]

In [7]:
# Check duplicate user_id again
df2[df2["user_id"].duplicated()]

user_id timestamp      group landing_page  converted
2893     773192   55:59.6  treatment     new_page          0
294478   759899   20:29.0  treatment     new_page          0

In [8]:
df2.drop_duplicates(subset="user_id",inplace=True)

/tmp/ipython-input-8-374184420.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2.drop_duplicates(subset="user_id",inplace=True)


In [9]:
df2.shape

(290585, 5)

In [10]:
# Conversion Rate
df_counts = pd.DataFrame({"Control_COUNT": df2[df2.group=="control"].converted.value_counts(),
              "Treatment_COUNT": df2[df2.group=="treatment"].converted.value_counts(),
              "Control_RATIO": df2[df2.group=="control"].converted.value_counts()/ len(df2[df2.group=="control"]),
              "Treatment_RATIO": df2[df2.group=="treatment"].converted.value_counts() / len(df2[df2.group=="treatment"])})
df_counts

Control_COUNT  Treatment_COUNT  Control_RATIO  Treatment_RATIO
converted                                                                
0                 127785           128047       0.879614         0.881193
1                  17489            17264       0.120386         0.118807

H0: conversion rate of Control verson = conversion rate of treatmentment

H1: conversion rate of Control verson != conversion rate of treatmentment

In [13]:
table = df_counts.iloc[:,0:2].to_numpy()

In [14]:
# chi-squared test
stat, p, dof, expected = stats.chi2_contingency(table)
print("fredoom of degree=%d"% dof)
print(expected)
# interpret test-statistic
prob = 0.95
critical = stats.chi2.ppf(prob, dof)
print('probability=%.3f, critical=%.3f, stat=%.3f' % (prob, critical, stat))
if abs(stat) >= critical:
    print('Dependent (reject H0)')
else:
    print('Independent (fail to reject H0)')
# interpret p-value
alpha = 1-prob
print('significance=%.3f, p=%.3f' % (alpha, p))
if p <= alpha:
    print('Dependent (reject H0)')
else:
    print('Independent (fail to reject H0)')

fredoom of degree=1
[[127899.7125385 127932.2874615]
 [ 17374.2874615  17378.7125385]]
probability=0.950, critical=3.841, stat=1.705
Independent (fail to reject H0)
significance=0.050, p=0.192
Independent (fail to reject H0)


In [15]:
# Fishers exact test on the data
odd_ratio, p_value = stats.fisher_exact(table, alternative="two-sided")
print("odd ratio is : " + str(odd_ratio))
print("p_value is : " + str(p_value))
if p_value <= alpha:
    print('Dependent (reject H0)')
else:
    print('Independent (fail to reject H0)')

odd ratio is : 0.9851149705891606
p_value is : 0.19047607131914918
Independent (fail to reject H0)
